core/server/master/src/main/java/alluxio/master/file/LostFileDetector.java

```diff
@@ -37,40 +43,64 @@ final class LostFileDetector implements HeartbeatExecutor {
+  public LostFileDetector(FileSystemMaster fileSystemMaster, BlockMaster blockMaster,
+      InodeTree inodeTree) {
     mFileSystemMaster = fileSystemMaster;
     mInodeTree = inodeTree;
+    mBlockMaster = blockMaster;
     MetricsSystem.registerCachedGaugeIfAbsent(MetricKey.MASTER_LOST_FILE_COUNT.getName(),
         () -> mFileSystemMaster.getLostFiles().size());
   }
 
   @Override
   public void heartbeat() throws InterruptedException {
-    for (long fileId : mFileSystemMaster.getLostFiles()) {
+    Iterator<Long> iter = mBlockMaster.getLostBlocksIterator();
+    Set<Long> lostFiles = new HashSet<>();
+    while (iter.hasNext()) {
+      long blockId = iter.next();
+      long containerId = BlockId.getContainerId(blockId);
+      long fileId = IdUtils.createFileId(containerId);
+      lostFiles.add(fileId);
+      iter.remove();
+    }
+    for (long fileId : lostFiles) {
       // Throw if interrupted.
       if (Thread.interrupted()) {
         throw new InterruptedException("LostFileDetector interrupted.");
       }
-      // update the state
-      try (JournalContext journalContext = mFileSystemMaster.createJournalContext();
-          LockedInodePath inodePath =
-              mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
+      boolean markAsLost = false;
+      try (LockedInodePath inodePath = mInodeTree.lockFullInodePath(fileId, LockPattern.READ)) {
         Inode inode = inodePath.getInode();
         if (inode.getPersistenceState() != PersistenceState.PERSISTED) {
-          mInodeTree.updateInode(journalContext, UpdateInodeEntry.newBuilder()
-              .setId(inode.getId())
-              .setPersistenceState(PersistenceState.LOST.name())
-              .build());
+          markAsLost = true;
         }
       } catch (FileDoesNotExistException e) {
         LOG.debug("Exception trying to get inode from inode tree", e);
-      } catch (UnavailableException e) {
-        LOG.warn("Journal is not available. Backing off. Error: {}", e.toString());
-        break;
+      }
+
+      if (markAsLost) {
+        // update the state
+        try (JournalContext journalContext = mFileSystemMaster.createJournalContext();
+             LockedInodePath inodePath =
+                 mInodeTree.lockFullInodePath(fileId, LockPattern.WRITE_INODE)) {
+          Inode inode = inodePath.getInode();
+          if (inode.getPersistenceState() != PersistenceState.PERSISTED) {
+            mInodeTree.updateInode(journalContext,
+                UpdateInodeEntry.newBuilder().setId(inode.getId())
+                    .setPersistenceState(PersistenceState.LOST.name()).build());
+            lostFiles.add(fileId);
+          }
+        } catch (FileDoesNotExistException e) {
+          LOG.debug("Exception trying to get inode from inode tree", e);
+        } catch (UnavailableException e) {
+          LOG.warn("Journal is not available. Backing off. Error: {}", e.toString());
```

Comment:
```suggestion
          LOG.warn("Failed to mark files LOST because the journal is not available. {} files are affected: {}", lostFiles.size(), lostFile, e);
```